Feature/onix workflow

tuanchien opened the pull request:
Workflow telescope that waits on ONIX telescope, aggregates into works and work families, and produces bigquery lookup tables for them.

- oaebu intermediate tables will be handled in a separate branch
- there's been no testing of externaltasksensor code. that's probably better left for integration testing.

jdddog commented on observatory-dags/observatory/dags/workflows/onix_workflow.py:
```diff
@@ -0,0 +1,314 @@
+        if bq_client is None:
+            self.client = Client()
+
+        # GCP parameters
+        self.workflow_dataset_id = "onix_workflow"
+        self.project_id = gcp_project_id
+        self.onix_dataset_id = "onix"
+        self.dataset_location = "us"
+        self.dataset_description = "ONIX workflow tables"
+
+        # ONIX release info
+        table_name, onix_release_date_str = self.get_latest_table_id()
+        self.onix_table_id = table_name
+        self.release_date = pendulum.parse(onix_release_date_str)
+        self.bucket_name = gcp_bucket_name
+        self.blob_path = os.path.join("onix_workflow", onix_release_date_str)
+
+        # Prepare filesystem
+        self.local_table_dir = self.blob_path
+        self.workslookup_filename = os.path.join(self.local_table_dir, "onix_workid_isbn.jsonl.gz")
+        self.workslookup_errors_filename = os.path.join(self.local_table_dir, "onix_workid_isbn_errors.jsonl.gz")
+        self.worksfamilylookup_filename = os.path.join(self.local_table_dir, "onix_workfamilyid_isbn.jsonl.gz")
+
+        Path(".", self.local_table_dir).mkdir(exist_ok=True, parents=True)
+
+        # Work (family) aggregation results
+        self.works = list()
+        self.works_families = list()
```
It would be good to put these into a release object similar to the other DAGs and create the release object as part of an Airflow task.

The reason is that, the OnixWorkflow is constructed, and therefore the `release_date` is fetched, when the DAG file is loaded, not after waiting for the ONIX telescope to finish (via the external task sensor).

jdddog commented on observatory-dags/observatory/dags/workflows/onix_workflow.py:
```diff
@@ -0,0 +1,314 @@
+
+        return None
+
+    def get_latest_table_id(self) -> Tuple[str, str]:
+        """Get the newest table name from the onix dataset.
+
+        :return: Latest table name and date string.
+        """
+
+        sql = f"SELECT table_name from {self.project_id}.{self.onix_dataset_id}.INFORMATION_SCHEMA.TABLES ORDER BY table_name DESC LIMIT 1"
+        df = self.client.query(sql).to_dataframe()
+        table_name = df["table_name"][0]
+        date_str = table_name[-8:]  # Last 8 digits is the YYYYMMDD date string
+        return table_name, date_str
```
There is already a function here that can be used to select the table shard dates for a particular table: https://github.com/The-Academic-Observatory/observatory-platform/blob/develop/observatory-dags/observatory/dags/workflows/doi.py#L50-L68

It could be moved into the telescope utilities and renamed into something that makes better sense. You can also specify an end date to search up to (so that you only include tables up to the release date of the workflow for instance, in the case where there might be future releases of a table, such as when back-filling).

I think you can assume that the ONIX tables will always be called `onix`, so that could just be hardcoded as a constant in the telescope.

jdddog reviewed (changes requested):
Hey Tuan, it looks good, just a couple of comments.

tuanchien commented:
- Refactored to use select_table_suffix, and put that function in gc_utils
- Using a release object now.  Not sure if this is particularly fit for purpose though since half of the methods in AbstractRelease are unused.
- Do all the in-memory work(family) aggregation in a single task to avoid needing to share info between tasks.
- Changed ISBN/GTIN in onix schema to string instead of integer.
- Added tasks for creating oaebu_intermediate tables that have work(family) ids added. This was initially intended for an independent branch.  There is a temporary hard coded method for populating the data partner info until we get a more long term solution in place later.

jdddog commented on observatory-dags/observatory/dags/workflows/onix_workflow.py:
```diff
@@ -0,0 +1,509 @@
+        :return: OnixWorkflowRelease object.
+        """
+
+        release_date = kwargs["execution_date"]
+        release = OnixWorkflowRelease(
+            dag_id=self.dag_id,
+            release_date=release_date,
+            gcp_project_id=self.gcp_project_id,
+            gcp_bucket_name=self.gcp_bucket_name,
+            onix_dataset_id=self.onix_dataset_id,
+            onix_table_id=self.onix_table_id,
+        )
```
The biggest thing that will need changing is refactoring the OnixWorkflow to process multiple releases. The OnixTelescope processes multiple releases for each DAG run (e.g. if multiple ONIX files have been deposited), therefore the OnixWorkflow should also be able to process multiple releases.

I'm thinking that a better way to communicate what releases need to be processed by the OnixWorkflow, is to pull the XCom published by the OnixTelescope for the same period. This XCom contains a list of release dates, e.g. `[{'release_date': <Pendulum [2021-04-20T00:00:00+00:00]>, 'file_name': '20210420_ANUPRESS_ONIX.xml'}]`. This means that you would not use `select_table_suffixes` further down, just the release dates pulled from XCom stored in the release object.

You can pull them like so:
```python
ti: TaskInstance = kwargs['ti']
records = ti.xcom_pull(
    dag_id='onix_telescope_name',
    key=OnixTelescope.RELEASE_INFO,
    task_ids='list_release_info',
    include_prior_dates=False)
```

jdddog reviewed (changes requested):
It is looking good Tuan, see the comments below for suggestions for changes.

tuanchien commented on observatory-dags/observatory/dags/workflows/onix_workflow.py:
```diff
@@ -0,0 +1,509 @@
+        )
+
+        # Wait on ONIX telescope to finish
+        self.add_sensor(telescope_sensor)
```
I was trying to avoid having a dependency on Observatory API classes by passing in a sensor pre-constructed. Would you prefer the org name gets passed to the telescope?

jdddog reviewed (changes requested):
Hey Tuan, that looks good, just a couple of minor suggestions and then I think we can merge it.
